kv/kvclient/kvcoord: TestWaiterOnRejectedCommit failed

kv/kvclient/kvcoord.TestWaiterOnRejectedCommit [failed](https://mesolite.cluster.engflow.com/invocations/default/7771edf4-73d6-4e4c-b19f-41c27d725f16?testReportRun=1&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZjbGllbnQva3Zjb29yZDprdmNvb3JkX3Rlc3Q=) on release-24.3.32-rc @ [c834642eaf4393c6c51d7aaa493cd2770cb00de9](https://github.com/cockroachdb/cockroach/commits/c834642eaf4393c6c51d7aaa493cd2770cb00de9):


```
=== RUN   TestWaiterOnRejectedCommit
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestWaiterOnRejectedCommit2637243324
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:139: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    integration_test.go:168: txn ID: 63647e2a-231b-45c8-be70-d9500956832e
    integration_test.go:88: EndTxn command ID: 7399b00fc7a328dc
    integration_test.go:103: injecting illegal lease index for 7399b00fc7a328dc
    integration_test.go:190: 
        	Error Trace:	pkg/kv/kvclient/kvcoord/integration_test.go:190
        	Error:      	An error is expected but got nil.
        	Test:       	TestWaiterOnRejectedCommit
        	Messages:   	expected injected error
    panic.go:626: -- test log scope end --
--- FAIL: TestWaiterOnRejectedCommit (4.87s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWaiterOnRejectedCommit.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63048

--- a/pkg/kv/kvclient/kvcoord/integration_test.go
+++ b/pkg/kv/kvclient/kvcoord/integration_test.go
@@ -52,8 +52,6 @@ func TestWaiterOnRejectedCommit(t *testing.T) {
 
 	// The txn id whose commit we're going to reject. A uuid.UUID.
 	var txnID atomic.Value
-	// The EndTxn proposal that we want to reject. A string.
-	var commitCmdID atomic.Value
 	readerBlocked := make(chan struct{}, 2)
 	// txnUpdate is signaled once the txn wait queue is updated for our
 	// transaction. Normally it only needs a buffer length of 1, but bugs that
@@ -62,53 +60,40 @@ func TestWaiterOnRejectedCommit(t *testing.T) {
 	txnUpdate := make(chan roachpb.TransactionStatus, 20)
 
 	illegalLeaseIndex := true
-	s, sqlDB, db := serverutils.StartServer(t, base.TestServerArgs{
+	s, _, db := serverutils.StartServer(t, base.TestServerArgs{
 		Knobs: base.TestingKnobs{
 			Store: &kvserver.StoreTestingKnobs{
 				DisableMergeQueue: true,
 				DisableSplitQueue: true,
-				TestingProposalFilter: func(args kvserverbase.ProposalFilterArgs) *kvpb.Error {
-					// We'll recognize the attempt to commit our transaction and store the
-					// respective command id.
-					ba := args.Req
-					etReq, ok := ba.GetArg(kvpb.EndTxn)
+				TestingApplyCalledTwiceFilter: func(args kvserverbase.ApplyFilterArgs) (int, *kvpb.Error) {
+					// Match on the transaction ID rather than the command ID,
+					// because reproposals get new command IDs.
+					if args.Req == nil {
+						return 0, nil
+					}
+					etReq, ok := args.Req.GetArg(kvpb.EndTxn)
 					if !ok {
-						return nil
+						return 0, nil
 					}
 					if !etReq.(*kvpb.EndTxnRequest).Commit {
-						return nil
+						return 0, nil
 					}
 					v := txnID.Load()
 					if v == nil {
-						return nil
-					}
-					if !ba.Txn.ID.Equal(v.(uuid.UUID)) {
-						return nil
-					}
-					t.Logf("EndTxn command ID: %v", args.CmdID)
-					commitCmdID.Store(args.CmdID)
-					return nil
-				},
-				TestingApplyCalledTwiceFilter: func(args kvserverbase.ApplyFilterArgs) (int, *kvpb.Error) {
-					// We'll trap the processing of the commit command and return an error
-					// for it.
-					v := commitCmdID.Load()
-					if v == nil {
 						return 0, nil
 					}
-					cmdID := v.(kvserverbase.CmdIDKey)
-					if args.CmdID == cmdID {
-						if illegalLeaseIndex {
-							illegalLeaseIndex = false
-							t.Logf("injecting illegal lease index for %v", cmdID)
-							// NB: 1 is ProposalRejectionIllegalLeaseIndex.
-							return 1, kvpb.NewErrorf("test injected err (illegal lease index)")
-						}
-						t.Logf("injecting permanent rejection for %v", cmdID)
-						// NB: 0 is ProposalRejectionPermanent.
-						return 0, kvpb.NewErrorf("test injected err")
+					if !args.Req.Txn.ID.Equal(v.(uuid.UUID)) {
+						return 0, nil
 					}
-					return 0, nil
+					if illegalLeaseIndex {
+						illegalLeaseIndex = false
+						t.Logf("injecting illegal lease index for %v", args.CmdID)
+						// NB: 1 is ProposalRejectionIllegalLeaseIndex.
+						return 1, kvpb.NewErrorf("test injected err (illegal lease index)")
+					}
+					t.Logf("injecting permanent rejection for %v", args.CmdID)
+					// NB: 0 is ProposalRejectionPermanent.
+					return 0, kvpb.NewErrorf("test injected err")
 				},
 				TxnWaitKnobs: txnwait.TestingKnobs{
 					OnPusherBlocked: func(ctx context.Context, push *kvpb.PushTxnRequest) {
@@ -136,16 +121,6 @@ func TestWaiterOnRejectedCommit(t *testing.T) {
 		},
 	})
 	defer s.Stopper().Stop(ctx)
-
-	// Disable parallel commits protocol. In some rare cases, we see two different
-	// command ids for the committing this transaction. The first one could be the
-	// one through parallel commits protocol and the second is using the normal
-	// commit protocol. There is a test race where TestingProposalFilter changes
-	// the commit command id from the first one (parallel commit) to the second
-	// one (normal commit). This causes the test to inject the error on the wrong
-	// command id. Disabling parallel commits avoids this test race.
-	_, err := sqlDB.Exec("SET CLUSTER SETTING kv.transaction.parallel_commits_enabled = 'false'")
-	require.NoError(t, err)
 
 	if _, _, err := s.SplitRange(roachpb.Key("b")); err != nil {
 		t.Fatal(err)

